lib/api/api_auth.go

```diff
@@ -82,33 +82,40 @@ func basicAuthAndSessionMiddleware(cookieName string, guiCfg config.GUIConfigura
 			}
 		}
 
-		l.Debugln("Sessionless HTTP request with authentication; this is expensive.")
-
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			unauthorized(w)
+		if username, ok := attemptBasicAuth(r, guiCfg, ldapCfg, evLogger); ok {
```

Comment:
The typical go way is to check for failure (`err != nil`, `!ok`), handle that (and likely return) and otherwise continue doing things after the condition. While it doesn't matter here, as both branches are short, I'd still like to follow that convention and invert the condition here, as the control flow is less surprising to the reader.

Reply:
I won't refuse to make that change, but I think that security checks in particular is a context where it may be better to invert that approach. The `!err`/`!ok` pattern checks for specific error conditions, aborts if any of them is present, and _otherwise proceeds_ along the happy path. But in security checks I think it's usually best to think the other way around: check for specific clearance conditions and allow them through, and _otherwise reject_ the request.

I think this is a better way to structure programs to be _fail-safe_ rather than brittle. If in future code revisions an early `return` goes missing, it's no disaster if we end up creating two sessions. But it _is_ a disaster if an early `return` goes missing and we end up letting an illegitimate user through.

If you still prefer `!err`/`!ok` pattern I'll change to that, but that's my rationale for structuring it like this.

Reply:
Nice, thanks for explaining - haven't thought about that before. As in all good as is then.

(Personal, unrelated note: Yeiii - found someone who gets review comments for what they are (I want them to be): Reviewer pointing out potential for improval/change with a rationale, reviewee (that a thing?) reacting with either explaining their view or doing the change. Way too often only the latter seems to be considered as an expected outcome, and thus things become unnecessarily defensive.).

Reply:
Thanks! (And yeiii agreed! :smile:)